---
title: "UFMG-BiocRNAseq"
author: "Iza"
date: "2025-01-17"
output: html_document
---

Suggested bibliography:
original bioc-rnaseq:https://carpentries-incubator.github.io/bioc-rnaseq/

## 1. Introduction to RNA-seq
Questions:
* What are the different choices to consider when planning an RNA-seq experiment?
* How does one process the raw fastq files to generate a table with read counts per gene and sample?
* Where does one find information about annotated genes for a given organism?
* What are the typical steps in an RNA-seq analysis?

###Iza:This we shoudl do through slides.
Questions:
*
*
*
## 2. Where to find and how to download experimental Data
The dataset is available at Gene Expression Omnibus (GEO), under the accession number GSE96870. 

###Iza: Open the GEO show what is GEO, how to download the data
```{r}
download.file(
    url = "https://github.com/carpentries-incubator/bioc-rnaseq/raw/main/episodes/data/GSE96870_counts_cerebellum.csv", 
    destfile = "data/GSE96870_counts_cerebellum.csv"
)
```

## Importing and annotating quantified data into R
Questions:
*How can one import quantified gene expression data into an object suitable for downstream statistical analysis in R?
*What types of gene identifiers are typically used, and how are mappings between them done?

Load packages
```{r}
suppressPackageStartupMessages({
    library(AnnotationDbi)
    library(org.Mm.eg.db)
    library(hgu95av2.db)
    library(SummarizedExperiment)
})
```

Read the counts data
```{r}
counts <- read.csv("data/GSE96870_counts_cerebellum.csv", 
                   row.names = 1)
dim(counts)

View(counts)
```

Read the sample annotations
```{r}
coldata <- read.csv("data/GSE96870_coldata_cerebellum.csv",
                    row.names = 1)
dim(coldata)

View(coldata)
```

Read annotations
```{r}
rowranges <- read.delim("data/GSE96870_rowranges.tsv", 
                        sep = "\t", 
                        colClasses = c(ENTREZID = "character"),
                        header = TRUE, 
                        quote = "", 
                        row.names = 5)
dim(rowranges)

table(rowranges$gbkey)
```

### Assemble a Summarized Experiment
####iza: Explain what is an S4 object maybe some slides ?
```{r}
all.equal(colnames(counts), rownames(coldata)) # samples
all.equal(rownames(counts), rownames(rowranges)) # genes

tempindex <- match(colnames(counts), rownames(coldata))
coldata <- coldata[tempindex, ]

# Check again:
all.equal(colnames(counts), rownames(coldata)) 
```

Check everything before assemblying
```{r}
# One final check:
stopifnot(rownames(rowranges) == rownames(counts), # features
          rownames(coldata) == colnames(counts)) # samples

se <- SummarizedExperiment(
    assays = list(counts = as.matrix(counts)),
    rowRanges = as(rowranges, "GRanges"),
    colData = coldata
)
```

Common mistakes
```{r}
bad1 <- SummarizedExperiment(
    assays = list(counts = as.matrix(counts)),
    rowRanges = as(rowranges, "GRanges"),
    colData = coldata[1:3,]
)

bad2 <- SummarizedExperiment(
  assays = list(counts = as.matrix(counts)),
  rowRanges = as(rowranges[c(2:nrow(rowranges), 1),], "GRanges"),
  colData = coldata
)
```


Access the data
```{r}
colData(se)
dim(colData(se))
head(rowData(se))
dim(rowData(se))

```

Edit teh metadata information
```{r}
# Make better sample IDs that show sex, time and mouse ID:

se$Label <- paste(se$sex, se$time, se$mouse, sep = "_")
se$Label

colnames(se) <- se$Label

# Our samples are not in order based on sex and time
se$Group <- paste(se$sex, se$time, sep = "_")
se$Group

se$Group <- factor(se$Group, levels = c("Female_Day0","Male_Day0", 
                                        "Female_Day4","Male_Day4",
                                        "Female_Day8","Male_Day8"))
se <- se[, order(se$Group)]
colData(se)
```

Save your SE
```{r}
saveRDS(se, "data/GSE96870_se.rds")
rm(se) # remove the object!
se <- readRDS("data/GSE96870_se.rds")
```

